rust/src/dns/dns.rs

```diff
@@ -701,17 +700,21 @@ pub extern "C" fn rs_dns_parse_request_tcp(_flow: *const core::Flow,
                                            input_len: u32,
                                            _data: *const std::os::raw::c_void,
                                            _flags: u8)
-                                           -> std::os::raw::c_int {
+                                           -> AppLayerResult {
     let state = cast_pointer!(state, DNSState);
     if input_len > 0 {
         if input != std::ptr::null_mut() {
             let buf = unsafe{
                 std::slice::from_raw_parts(input, input_len as usize)};
-            return state.parse_request_tcp(buf) as std::os::raw::c_int;
+            if state.parse_request_tcp(buf) >= 0 {
+                return AppLayerResult::ok();
+            } else {
+                return AppLayerResult::err();
+            }
         }
         state.request_gap(input_len);
     }
-    return 0;
+    AppLayerResult::ok()
```

Comment:
This does functionally change the behaviour of this parser on error. Currently in master, -1 can never be returned, so it will continue trying to parse this flow even after a failure. This change will ripply the failure back to the app-layer. I'm not sure whats best, but I'm not sure if we change the current behaviour in this PR.

Reply:
I see the change, but not here. Here the `AppLayerResult::err()` condition seems to be unreachable. For UDP, I see it clearly.

Reply:
Hmm no for UDP we already did pass down failure.

Reply:
So do you see a patch for the TCP parser to reach the error return path?

Reply:
You're right, err is never returned as we never return < 0.

Reply:
Fixing it up anyway.